extended/src/main/java/io/kubernetes/client/extended/controller/WorkQueueResourceEventHandler.java

```diff
@@ -0,0 +1,116 @@
+package io.kubernetes.client.extended.controller;
+
+import io.kubernetes.client.extended.controller.reconciler.Reconciler;
+import io.kubernetes.client.extended.workqueue.WorkQueue;
+import io.kubernetes.client.informer.ResourceEventHandler;
+import java.util.function.BiPredicate;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * An event handler plumbs work-queue into a controller.
+ *
+ * @param <ApiType> the type parameter
+ */
+public class WorkQueueResourceEventHandler<ApiType> implements ResourceEventHandler<ApiType> {
+
+  /**
+   * Instantiates a new Work queue resource event handler.
+   *
+   * @param workQueue the work queue
+   * @param workKeyGenerator the work key generator
+   */
+  public WorkQueueResourceEventHandler(
+      WorkQueue<Reconciler.Request> workQueue,
+      Function<ApiType, Reconciler.Request> workKeyGenerator) {
+    this.workQueue = workQueue;
+    this.workKeyGenerator = workKeyGenerator;
+  }
+
+  private final WorkQueue<Reconciler.Request> workQueue;
+
+  private final Function<ApiType, Reconciler.Request> workKeyGenerator;
+
+  private Predicate<ApiType> onAddFilterPredicate;
+  private BiPredicate<ApiType, ApiType> onUpdateFilterPredicate;
+  private BiPredicate<ApiType, Boolean> onDeleteFilterPredicate;
+
+  /**
+   * Gets on add filter predicate.
+   *
+   * @return the on add filter predicate
+   */
+  public Predicate<ApiType> getOnAddFilterPredicate() {
+    return onAddFilterPredicate;
+  }
+
+  /**
+   * Sets on add filter predicate.
+   *
+   * @param onAddFilterPredicate the on add filter predicate
+   */
+  public void setOnAddFilterPredicate(Predicate<ApiType> onAddFilterPredicate) {
+    this.onAddFilterPredicate = onAddFilterPredicate;
+  }
+
+  /**
+   * Gets on update filter predicate.
+   *
+   * @return the on update filter predicate
+   */
+  public BiPredicate<ApiType, ApiType> getOnUpdateFilterPredicate() {
+    return onUpdateFilterPredicate;
+  }
+
+  /**
+   * Sets on update filter predicate.
+   *
+   * @param onUpdateFilterPredicate the on update filter predicate
+   */
+  public void setOnUpdateFilterPredicate(BiPredicate<ApiType, ApiType> onUpdateFilterPredicate) {
+    this.onUpdateFilterPredicate = onUpdateFilterPredicate;
+  }
+
+  /**
+   * Gets on delete filter predicate.
+   *
+   * @return the on delete filter predicate
+   */
+  public BiPredicate<ApiType, Boolean> getOnDeleteFilterPredicate() {
+    return onDeleteFilterPredicate;
+  }
+
+  /**
+   * Sets on delete filter predicate.
+   *
+   * @param onDeleteFilterPredicate the on delete filter predicate
+   */
+  public void setOnDeleteFilterPredicate(BiPredicate<ApiType, Boolean> onDeleteFilterPredicate) {
+    this.onDeleteFilterPredicate = onDeleteFilterPredicate;
+  }
+
+  @Override
+  public void onAdd(ApiType obj) {
+    if (onAddFilterPredicate != null && onAddFilterPredicate.negate().test(obj)) {
```

Comment:
Personally this feels cleaner as:

```java
if (onAddFilterPredicate == null || onAddFilterPredicate.test(obj)) {
   workQueue.add(...);
}
```

but it's your call.

Reply:
yours looks clearer. updated.